cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/QueryPlanTest.scala

```diff
@@ -441,16 +836,287 @@ class QueryPlanTest extends DocumentingTestBase with SoftReset {
+          |ON CREATE SET p.exists = true""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("AntiConditionalApply"))
+    )
+  }
+
+  @Test def assertSameNode() {
+    profileQuery(
+      title = "Assert Same Node",
+      text =
+        """The `AssertSameNode` operator is used to ensure that no uniqueness constraints are violated.
+          |The example looks for the existence of a team with the supplied name and id, and if one does not exist,
+          |it will be created. Owing to the existence of two uniqueness constraints
+          |on `:Team(name)` and `:Team(id)`, any node that would be found by the `UniqueIndexSeek`
+          |must be the very same node, or the constraints would be violated.
+        """.stripMargin,
+      queryText =
+        """MERGE (t:Team {name: 'Engineering', id: 42})""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("AssertSameNode"))
+    )
+  }
+
+  @Test def nodeHashJoin() {
+    executePreparationQueries(
+      List(
+        """MATCH (london:Location {name: 'London'}), (person:Person {name: 'Pontus'})
+          FOREACH(x in range(0, 250) |
+            CREATE (person) -[: WORKS_IN] ->(london)
+            )""".stripMargin
+      )
+    )
+    profileQuery(
+      title = "Node Hash Join",
+      text =
+        """Using a hash table, the `NodeHashJoin` operator joins the input coming from the left child operator with the input coming from the right child operator.
+          |`NodeHashJoin` only gets planned for larger cardinalities; for smaller cardinalities, `Expand` is used instead.""".stripMargin,
+      queryText =
+        """MATCH (andy:Person {name:'Andreas'})-[:WORKS_IN]->(loc)<-[:WORKS_IN]-(matt:Person {name:'Mattis'})
+          |RETURN loc.name""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("NodeHashJoin"))
+    )
+  }
+
+  @Test def triadic() {
+    profileQuery(
+      title = "Triadic",
+      text =
+        """The `Triadic` operator is used to solve triangular queries, such as the very
+          |common 'find my friend-of-friends that are not already my friend'.
+          |It does so by putting all the friends into a set, and uses the set to check if the
+          |friend-of-friends are already connected to me.
+          |The example finds the names of all friends of my friends that are not already my friends.""".stripMargin,
+      queryText =
+        """MATCH (me:Person)-[:FRIENDS_WITH]-()-[:FRIENDS_WITH]-(other)
+          |WHERE NOT (me)-[:FRIENDS_WITH]-(other)
+          |RETURN other.name""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("Triadic"))
+    )
+  }
+
+  @Test def foreach() {
+    profileQuery(
+      title = "Foreach",
+      text =
+        """The `Foreach` operator xxxx.""".stripMargin,
```

Comment:
The `Foreach` operator executes a nested loop between the LHS and RHS - just like `Apply` it takes a row from the LHS and feeds it to the operator tree on the RHS using the `Argument` operator. It will pull all the results from the RHS, but will just discard them, just yield all the rows coming in from the LHS.

Reply:
👍